File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.java
class Solution {
    int helper(int nums[],int i,int flag,int dp[][]){
        if(i >= nums.length){
            return 0;
        }
        if(dp[i][flag] != -1) return dp[i][flag];
        
        if(flag == 0){
            int opt1 = helper(nums,i+1,1,dp) - nums[i];
            int opt2 = helper(nums,i+1,0,dp);
            return dp[i][flag] = Math.max(opt1,opt2);
        }
        else {
            int opt1 = helper(nums,i+1,0,dp) + nums[i];
            int opt2 = helper(nums,i+1,1,dp);
            return dp[i][flag] = Math.max(opt1,opt2);
        }
    }
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int dp[][] = new int[n][2];
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
        }
        return helper(prices,0,0,dp);
    }
}

File: 122-best-time-to-buy-and-sell-stock-ii/122-best-time-to-buy-and-sell-stock-ii.cpp
class Solution {
public:
//     int maxProfit(vector<int>& prices) {
        
//     }
    
    int helper(int i, int b, vector<int> &A,vector<vector<int>> &dp){
        if(i==A.size()) return 0;
        int a=0;

            if(dp[i][b] != -1) return dp[i][b];
        if(b==0){
            int x=helper(i+1,0,A,dp);
            int y=-A[i]+helper(i+1,1,A,dp);
            a=max(x,y);
        }
        else{
            int x= A[i] + helper(i+1,0,A,dp);
            int y= helper(i+1,1,A,dp);
            a=max(x,y);
        }
        return dp[i][b] = a;
    }
    int maxProfit(vector<int> &A) {
        if(A.size()<=1) return 0; 
        vector<vector<int>> dp(A.size(),vector<int>(2,-1));
        int n=A.size(), b=0;
        return helper(0,b,A,dp);
    }
};